### ISSUE TITLE
feature: warn when calling async functions without an await (no floating promises)

### ISSUE THREAD

**Issue by bigman73** [xjamundx/eslint-plugin-promise#151]
Forgetting to put an await before calling an async function can have very bad outcomes and is quite hard to notice

TSLint has a rule named no-floating-promises -https://palantir.github.io/tslint/rules/no-floating-promises/

Can you implement a similar rule in this ESLint plugin?

**Comment by david-eighmey** [comment]
I can take a look at this.

**Comment by macklinu** [comment]
Just to confirm, we are looking to warn about not awaiting an async function like this?

```js
async function myAsyncFunction() {
  // ...
}

// ok, because of .then()
myAsyncFunction().then()

// ok, because of .catch()
myAsyncFunction().catch()

// ok, because awaited inside another async function
;(async () => {
  let value = await myAsyncFunction()
})

// not ok, because not awaited
;(async () => {
  let value = myAsyncFunction()
})
```

If you have other code examples that could help us catch all possible cases, please comment with more! Also, if I'm misunderstanding what syntax this rule would warn about, please let me know too - thanks. 👍

**Comment by bigman73** [comment]
Yes

Simply put, an async function is called without an await in front of it, and as a result it runs in the background

**Comment by zone117x** [comment]
Surprised to find this still unsupported. This is a common cause of bug that are difficult to track down.

**Comment by xjamundx** [comment]
You can't really do this in eslint, because it wouldn't know if a function is async unelss it was defined in that same file

**Comment by zone117x** [comment]
Do you mean eslint can't infer if a value is a Promise? I'm obviously to eslint's typing inference capabilities.

**Comment by xjamundx** [comment]
Yeah it has no typing inference AFAIK. Think of it like this:

```
// car.js
export async function start() { /* ... */ }

// index.js
import { start } from './car'

async function main()  {
    start(); // eslint has no idea this should be awaited
}
```

Eslint could figure it out if they were defined in the same file though, but my strong feeling here is that a type system should be able to help with these problems and it's the way I'd recommend solving them.

**Comment by zone117x** [comment]
Oh I see. Another argument for using typescript.
Thanks for the info.

**Comment by vladimiry** [comment]
You could try `@typescript-eslint/no-floating-promises` rule if you go with TypeScript.

**Comment by bigman73** [comment]
thanks, but I don't use typescript.

**Comment by sk-** [comment]
@bigman73 have you looked at https://eslint.org/docs/rules/require-await

**Comment by bigman73** [comment]
Thanks, looks like a solution, I'll test it

**Comment by bigman73** [comment]
UPDATE: It doesn't work. The rule warns when there is an async definition but no usage of await, but it doesn't warn against a function that is async and no await precedes the call
